### ISSUE TITLE
Reduce confirmations when deleting branch

### ISSUE THREAD

**Issue by RussKie** [gitextensions/gitextensions#5771]
**Do you want to request a *feature* or report a *bug*?**
Feature, usability issue
This is one of my pet hates, which often sees me retreating to bash

**What is the current behavior?**
Whenever a user attempts to delete a local branch, it takes an awful lot of clicks to achieve the desired.

If a branch hasn't been merged into the current trunk, unless the user ticks "force delete", there will be a message box:
![image](https://user-images.githubusercontent.com/4403806/48458586-12aeb280-e81b-11e8-9586-ee0f90f5b84a.png)
So the user will tick the "force delete" and presented with a confirmation:
![image](https://user-images.githubusercontent.com/4403806/48458608-2c4ffa00-e81b-11e8-930e-241e3193eef6.png)

This is a nuisance. Any local branch should be easy to delete, in event of an error they can be easily restored via git-reflog.

**What is the expected behavior?**
We should minimise a number of confirmations.

I propose to
* update "Are you sure" confirmation with a checkbox "Do not ask again"
* add a setting to _Settings > Advanced > Confirmations_ that control whether a user is prompted to confirm local branch deletion
* "Are you sure" confirmation is only to be shown for branches unmerged into the current trunk
* replace "Force delete" checkbox with a checkbox similar to the one on the "Remote Branch Delete" form, that serves as a confirmation of intent and enables the "Delete" button.
* update the help text on the form to something like the following:
  ```
  When you delete a branch the commits can get lost because nothing points to them.
  In event where you deleted a wrong branch, you can restore it with git-reflog.
  ```

**Environment you encounter the issue:**
all

**Did this work in previous version of GitExtensions (which)?**
no

Relates to #5027

**Comment by gerhardol** [comment]
> Perhaps we add a confirmation override for this (a "Don't ask me this again" checkbox).

agree

**Comment by RussKie** [comment]
Thank you for the feedback, I'll update the wording.

**Comment by RussKie** [comment]
Updated, please have a look

**Comment by gerhardol** [comment]
I propose to
* update "Are you sure" confirmation with a checkbox "Do not ask again"
* add a setting to _Settings > Advanced > Confirmations_ that control whether a user is prompted to confirm local branch deletion
~~* "Are you sure" confirmation is only to be shown for branches unmerged into the current trunk~~
~~* replace "Force delete" checkbox with a checkbox similar to the one on the "Remote Branch Delete" form, that serves as a confirmation of intent and enables the "Delete" button.~~
* update the help text on the form to something like the following:
  ```
  When you delete a branch the commits can get lost because nothing points to them.
  In event where you deleted a wrong branch, you can restore it with git-reflog.
  ```

--
 * Keep confirmation for all, unless you suppress confirmation, consistent with other GE checks
 * Keep --force, closer to Git command functionality (and documentation). If deleting unmerged, offer to override from the popup (compare with FormCommit Force push).

**Comment by vbjay** [comment]
Instead of got reflog point out you can view the reflog by changing view
option.

**Comment by amaiorano** [comment]
While we're at it, can we add an option to also delete the local branch when deleting the remote one? This is also a nuisance when working with feature branches after they've been merged.

**Comment by vbjay** [comment]
No. Already discussed.

**Comment by vbjay** [comment]
We have prune for when the remote is removed and you still have a tracking
branch.

You can delete a tracking branch and deuere from a remote. We don't want
all references to be deleted in one shot. Yes we have the reflog but we
want an extra layer of you having to manually delete a local branch as a
separate action.

**Comment by amaiorano** [comment]
I understand, but even with prune, this means I have to perform two separate actions to delete a single feature branch. I don't typically delete a bunch of remote branches at once, so pruning doesn't save me much. I see this in the same spirit as offering the "force" checkbox in the push dialog - I know this is actually an argument to "git push", but the point is that you want to avoid excessive actions. When deleting a remote branch, you almost always want to delete the local tracking branch.

Anyway, this is probably the wrong place to discuss this, as you said there was already a discussion about it.

**Comment by vbjay** [comment]
So deal with it.

**Comment by vbjay** [comment]
There are some things we did on purpose. We support newbs to advanced
users. There are some things we do to help the newbs with learning git and
making it so they don't lose their work.

**Comment by amaiorano** [comment]
> So deal with it.

Salty much? There's no need to respond that way.

**Comment by vbjay** [comment]
Let me also point you to the remove obsolete branches plugin.

**Comment by RussKie** [comment]
Jay, please stay respectful. Not everyone has a full history (or good
memory).

**Comment by ivangrek** [comment]
I think in this case the user should not worry about the features of the command interface (how it is done), but it would be a good decision to warn about the consequences.

We need to do what the user expects - delete the branch.

![image](https://user-images.githubusercontent.com/111697806/185990761-56ac038b-9005-48f7-883d-70e8aa609b98.png)

update the help text on the form to something like the following:

> When you delete a branch the commits can get lost because nothing points to them.
> In event where you deleted a wrong branch, you can restore it with git-reflog.


![image](https://user-images.githubusercontent.com/111697806/185992419-cb52d198-4ec1-411f-847a-b8b21ab70bd9.png)

And remove the confirmation dialog.


_If necessary, we could inform which command will be executed (this is also suitable for other cases, in other dialogs)._

`The command will be executed: git branch -D test`

![image](https://user-images.githubusercontent.com/111697806/185996250-a4a0fbe0-6b93-41d9-a591-b70765eca8fa.png)

**Comment by mstv** [comment]
> We need to do what the user expects - delete the branch.

👍

The check for "merged in HEAD" feels very arbitrary to me because it depends on the current checkout. And it delays the popup significantly (takes seconds).
I'd like to have (an option for advanced users which) always automatically force-deletes the selected branch - not checking the merged state before.
One popup should suffice.

In addition, I support  @amaiorano's proposal to be able to delete a (feature) branch from the local and the remote repo.

GE hides the git commands used behind. I do see a necessity to show a preview (which might become hard to maintain / to keep in sync).

**Comment by RussKie** [comment]
> `The command will be executed: git branch -D test`

That's an interesting proposal, but I don't think it's viable. First,  I don't believe a significant portion of our users care about the commands run. Secondly, we'd have to update all the dialogs, which is significant challenge (from both engineering and UX perspectives). Lastly, quite often we run multiple commands in parallel or sequence... 

A more viable option, in my view, is to make sure each dialog has `[?]` button, which takes the user the to the docs, that explain in details what the dialog does, what each option means, etc.

**Comment by ivangrek** [comment]
> In addition, I support @amaiorano's proposal to be able to delete a (feature) branch from the local and the remote repo.

Is it a proposal to make a single window with the ability to select all kinds of branches, or is it a proposal to add a function to delete all pointers (branches)?


It's implemented.
The question remains about the need for a checkmark when deleting a remote branch.

![image](https://user-images.githubusercontent.com/111697806/187077275-0308a582-6e9e-435b-9822-4a092aa8e392.png)

**Comment by mstv** [comment]
It does not work that way - at least not in v4.0.
The checkbox does not appear unless it is a remote branch. Then one is forced to tick the checkbox in order to proceed. :( The local branch will not be deleted.

**Comment by vbjay** [comment]
Protecting remote branch deletion is the idea. Being able to recover local
is a lot easier with reflog. You also affect others using repo when
deleting a remote branch. So preventing oops when deleting a remote branch
is the idea.

**Comment by ivangrek** [comment]
> Protecting remote branch deletion is the idea. Being able to recover local
is a lot easier with reflog. You also affect others using repo when
deleting a remote branch. So preventing oops when deleting a remote branch
is the idea.

For the local branch, protection is a warning. Why not do it the same way?

**Proposal.**

[Delete branch...] -> Open the delete branches dialog (for local and for remote) -> Select branches -> Show confirmation -> Delete selected branches.

[Delete branch...] -> [Specific branch] -> Show confirmation -> Immediately delete the selected branch.

**Comment by vbjay** [comment]
Because local is recoverable with reflog. You can recreate it. Remote
branches could be not in reflog.

Think of local and remote as two sandboxes. You can play in your sandbox
all you want and even recover since local is in reflog  but as soon as you
put your shovel in my sandbox (remote) there needs to be a guard from just
it happening.

**Comment by vbjay** [comment]
Answering the arbitrary comment. The head check verifies all commits are in
reflog of local since saying is merged with head is equivalent to saying
are call commits of remote in head. So if all megfed in head then it
confirms that if you delete the remote, you can at least recreate using
reflog. If you delete the remote and it is not merged, it is likely you
will lose commits on the remote repo and not be recoverable.

**Comment by mstv** [comment]
It hinders me to delete a merged feature branch if am working on the next feature branch, which has not been rebased yet...

**Comment by vbjay** [comment]
@mstv so do the deleting as last step before switching. Or use obsolete branches plugin to do a not head but specific branch check of merged to then chose the ones you want to remove. Since master/main are just branches and have no real special status in git as any other branch doesn't, head is as good as you can get. Also look at using git flow.

**Comment by gerhardol** [comment]
I prefer to to do it the with Git interface: Expose the force delete to the user
Alternatively, try delete first and ask if the user wants to retry with force if it fails, similar to push from Commit (where you also can override to set force by default)

**Comment by vbjay** [comment]
@mstv 
![image](https://user-images.githubusercontent.com/3629489/187084231-0869f54a-d8e0-4770-802c-eedd1c9dd86c.png)
I made a local branch from net7.0 branch and created arbitrary local branch from a merged branch earlier so one not merged and one merged.   

I have master checked out.  Not hard to switch out HEAD in the dialog to some other branch name to force master or other branch.  Can also check the checkbox to include unmerged.  Cleanup isn't that hard.  
![image](https://user-images.githubusercontent.com/3629489/187084420-45c20cac-f2cb-4038-9413-89cf377173c5.png)

You can even use
![image](https://user-images.githubusercontent.com/3629489/187084528-d27f4349-bbbe-41ef-af67-ce16cc0dd8ef.png)
which provides remote to local are we current and provides the same kind of checking but visual vs commit checking in the code.  





I prefer that when a remote branch is involved that a "the commits in y branch that is not in this repo could be lost" something is checked or alerted at least.  If I am using a dialog that is expressly for deleting REMOTE branches that something of this sort is used to either check the specific remote branch for do I have these commits locally or something like 
![image](https://user-images.githubusercontent.com/3629489/187084697-479f8bb8-5082-4577-b2ce-2eec31834157.png)
 that says there be potholes in the road you are going down.  

The one thing we cannot do is just sure goodbye remote repo/branchX blindly without some kind of guard.  Force delete is fine if I have to assert that through an action.  I would rather have something be annoying and alert me to the dangers than losing commits because is scrolled mouse wheel or whatever and selected the wrong branch and hit go.  Seeing a you are potentially deleting someone else's/central repo ( whatever flow determines what a remote is ) copy of a branch.  Local branches are FINE to delete because of the reflog.  Remote could be only remaining copy of branch with potential of history being lost.  

The force override where it checked if merged when doing remote made sense.  Because HEAD could be a dangling commit or other branch, I see the issue but you are not guaranteed to have a local tracking branch to compare merged state of a remote branch.  Maybe instead of delete local branch and also delete remote x copy of branch, we flip it to the other side.  Delete remote and do you want to delete local too?

Delete remote x branch and include an optional remove local  copy too if one exists.  If 

- No local copy then warn about losing history --  This one checks for merged state and warns 
- If deleting both at same time  then warn about the danger of losing history but mention reflog as possible recovery also where the tracking vs local are determines if reflog can recover or not.  If remote was pushed to but my local is behind...  I can lose history.

Yes exposing force delete option is a good idea.  HOWEVER, it needs to be opt in and never remembered and a there be dragons info provided that warns what that means.

**Comment by mstv** [comment]
Thank you for the explanation, Jay. Fair points - perhaps also for advanced users.
One remains unclear yet. I use to download the full history of a remote. So my local repo should be a complete copy of the remote. And if I delete a remote branch, I can recover it at least using `Recover lost objects`. It works well.

So, I still want to have an advanced mode which force-deletes by default without prior check whether the branch is merged to HEAD (which remains arbitrary regardless of that I understand that it is the best git can do).
I agree the `Delete remote branch` popup should not be confirmed by just `Enter`. Though `Alt+R`, `Alt+D` is a way too much. After ticking the checkbox, the `Delete` button should react on `Enter`, too, i.e. be the default button.

(Switching HEAD for being able to delete a branch is too heavy. I must avoid unnecessary rebuilds as far as possible.)

**Comment by vbjay** [comment]
@mstv You get that lost objects recoverability till your copy has commits A-F and something pushes commits G-X commits.  So removing a branch on a remote is never guaranteed to be completely copied to your repo unless you just did a fetch all.   This is why I say removing remote branches needs to be guarded.  Someone could have pushed G-X in the mean time you did other work and never pulled those commits.  The user that pushed these commits could drop their repo.   So only place that has G-X is your remote.  You remove that branch.  Basically I could have my sandcastle in my sandbox almost complete and have a turrets branch.  Your sandbox( clone of repo)  has a remote to my sandbox.  Unless you do a fetch all right before attempting this removal, you can't say remote x hasn't moved forward.  

Advanced behavior is fine and when enabled, could eliminate guards and warnings.  An I know what I am doing mode.  However, enabling such a mode should pop a warning describing what gets yanked.

**Comment by pmiossec** [comment]
(Late in the discussion, sorry...)

> The check for "merged in HEAD" feels very arbitrary to me because it depends on the current checkout.

I find this check also strange even if I also like the idea of GitExtensions preventing me to do a mistake (or do something that I will have difficulties to notice that I made a mistake or that will be time consuming to fix -- and even worst for co-workers not used to git--)...

So, **for local branch**,  because we are taking a lot of relying on the reflog (and if I remember well that's already the content of the label), it may be interesting to replace the check done by calling the command `git reflog exists <ref>` that checks that the commit is in the reflog.

Warn: Some tests should be done to see if it covers all our use cases (especially if it will remove the reflog of the branch because you are about to delete it so the commit won't be in the reflog after 😕) 

But if it is, maybe we could think about force delete by default if the commit is in the reflog because the user will be able to recover.
And if it is not, warn the user that commits will be lost which won't be recoverable and ask for a confirmation only in this case.

**Comment by vbjay** [comment]
I would want to at least see if a fetch of that branch is possible. I know
people are on planes or offline and can't always fetch.  If fetched then
could delete remote branch with knowledge it is at least recoverable. I
like the reflog exists check. Which prevents the head issues.

**Comment by amaiorano** [comment]
Hello, it's been a while! Reading through the latest conversation, one idea I had that might help is to make it so that Shift+Delete brings up delete dialog with "Force" already selected. This would be similar to how Shift+Delete in Explorer will delete a file without putting it in the Recycle bin. We could also enable the checkbox for deleting remote branches if Shift is pressed. 

Finally, if we do add an option to "Also delete local branch" when trying to delete a remote branch, this could also be checked if Shift is pressed. Basically, the Shift key is the advanced/power user feature. You opted into it when pressing Delete.

**Comment by vbjay** [comment]
Sounds good. I like it. The opt in is key.

**Comment by pmiossec** [comment]
> So, **for local branch**, because we are taking a lot of relying on the reflog (and if I remember well that's already the content of the label), it may be interesting to replace the check done by calling the command `git reflog exists <ref>` that checks that the commit is in the reflog.

Forget what I wrote, I misread the documentation. If we want to do that, we  should check ourself if the sha1 of the commit we want to delete is part of the commit hashes returned by the command `git reflog show HEAD --pretty=format:"%H"`

Thinking about it again, it will be surely a lot better than the current implemented test but won't cover all the use cases, especially if we use it also for remote deletion. Maybe used with the current HEAD control...

**Comment by vbjay** [comment]
@pmiossec  I like where you are going but the problem I see is this.  

You fetch remote X at  SHA X for branch X 10 days ago.  You don't fetch any more to get current.  You then think I should delete X branch on the remote.  Your check of is SHA X is only local understanding of where Remote X Branch X is pointing to.  It has no tie in to what the remote really is at unless you also do a current fetch.

**Comment by vbjay** [comment]
Could use 

```bash
 git reflog show --date=iso -n 1 upstream/master -z | grep -oP \@\{\(.*\)\}|sed -n "s/^.*@{\s*\(.*\)}/\1/p"

```
![image](https://user-images.githubusercontent.com/3629489/188054901-1444e267-4665-45ba-ac85-60b7a0a072cb.png)

replacing upstream/master with whatever branch  you are looking for.  That would get you when was it pulled in and hopefully nobody is actually committing on remote branch refs and are instead using tracking references 

That couk dtell us at least how long ago it was updated and make a guess on is upstream/master in our repo is likely pointing to what upstream/master is  pointing to in the upstream repo.  If stale add  extra protections and suggest fetching or whatever.

**Comment by pmiossec** [comment]
> @pmiossec I like where you are going but the problem I see is this.

In fact, to be clear on my point of view, I definitively think that deleting local branch could be improved (we should find a clever way to bother less the user --so maybe ask less confirmation-- and at least find a solution to #10122 ).

But I share your point on view with handling remote branches. We can really mess things when deleting remotes and it will be very difficult, especially for beginners, to fix it and some cases can't even be fixed alone and you will have to rely on other users repository (which is a big problem for me). **So I'm perfectly happy with how it works now for remote branches and I think we should not change the current workflow**

**Comment by vbjay** [comment]
@pmiossec 

 The tests I can see for remote are

- Was last fetch/pull too stale( will need to be defined as what stale means.  Is 5 minutes, 2 hours.... 🤷 ) See https://github.com/gitextensions/gitextensions/issues/5771#issuecomment-1235035647
- If not stale then is commit in reflog contained in a local branch( except the local tracking branch for the remote branch) instead of HEAD.  That answers the question correctly.  

After that we can either delete or say you must force delete because either stale or not in another branch than tracking branch.  

Like if you were to delete origin/pr1234 and the sha was aabb1234, then is in any local branch except local branch pr1234.  Understanding that pr1234 was the tracking branch for that remote branch.
